src/cursor/common.rs

```diff
@@ -57,23 +55,81 @@ where
     ) -> Self {
         let exhausted = spec.id() == 0;
         Self {
-            exhausted,
             client,
             provider: get_more_provider,
-            buffer: spec.initial_buffer,
-            post_batch_resume_token: None,
             info: spec.info,
-            pinned_connection,
+            state: Some(CursorState {
+                buffer: CursorBuffer::new(spec.initial_buffer),
+                exhausted,
+                post_batch_resume_token: None,
+                pinned_connection,
+            }),
             _phantom: Default::default(),
         }
     }
 
-    pub(super) fn take_buffer(&mut self) -> VecDeque<RawDocumentBuf> {
-        std::mem::take(&mut self.buffer)
+    pub(super) fn from_state(
+        state: CursorState,
+        client: Client,
+        info: CursorInformation,
+        provider: P,
+    ) -> Self {
+        Self {
+            provider,
+            client,
+            info,
+            _phantom: Default::default(),
+            state: state.into(),
+        }
+    }
+
+    pub(super) fn current(&self) -> Option<&RawDocument> {
+        self.state.as_ref().unwrap().buffer.current()
+    }
+
+    fn state_mut(&mut self) -> &mut CursorState {
+        self.state.as_mut().unwrap()
+    }
+
+    fn state(&self) -> &CursorState {
+        self.state.as_ref().unwrap()
+    }
+
+    /// Advance the cursor forward to the next document.
+    /// If there are no documents cached locally, perform getMores until
+    /// the cursor is exhausted or a result/error has been received.
+    pub(super) async fn advance(&mut self) -> Result<bool> {
+        loop {
+            self.state_mut().buffer.advance();
+
+            if !self.state().buffer.is_empty() {
+                break;
+            }
+
+            // if moving the offset puts us at the end of the buffer, perform another
+            // getMore if the cursor is still alive.
+
+            if self.state().exhausted {
+                return Ok(false);
+            }
+
+            let client = self.client.clone();
+            let spec = self.info.clone();
+            let pin = self.state().pinned_connection.replicate();
+
+            let result = self.provider.execute(spec, client, pin).await;
+            self.handle_get_more_result(result)?;
+        }
+
+        Ok(true)
+    }
+
+    pub(super) fn take_state(&mut self) -> CursorState {
+        self.state.take().unwrap()
     }
 
     pub(super) fn is_exhausted(&self) -> bool {
-        self.exhausted
+        self.state.as_ref().unwrap().exhausted
```

Comment:
For this and similar calls, I think it would be slightly better to centralize the `unwrap` in the `state()` accessor and write this as `self.state().exhausted`.

Reply:
Updated. I had to leave the one on line 245 in to satisfy the borrow checker, not sure why though.